rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+
+### What other approaches have been considered and why did you not choose them?
+
+The primary alternative implementation that's been discussed is to have spans be
+event-focused rather than component-focused. This would involve stashing a root
+span instance in a Vector event's metadata, and creating child spans to
+represent each stage of its processing. This would provide an end-to-end view of
+an event's journey through Vector, showing where the event itself spent time as
+opposed to where each component is spending time. While this would produce some
+interesting data, it would be significantly more complex and manual to
+implement, and it's not clear that the data would actually be more useful in
+addressing performance issues.
+
+Another alternative would be to avoid the coupling between tracing spans and
+logs, and use a library other than `tracing` for this functionality. This may
+provide some benefit in reducing confusion about log levels, accumulating events
+on spans, etc, but has the downside of requiring the introduction of a second
+system for instrumentation. It would also be rather confusing to use the
+`tracing` crate, but not for tracing, especially since it is one of the most
+widely used and well-engineered crates for this purpose.
+
+### How about not doing this at all?
+
+As discussed in the Rationale section, not doing this at all would leave us with
+a meaningful gap in our observability. There would be no immediate dire
+consequences, but we would expect to continue sinking more engineering time into
+debugging customer issues than we would if this data were easily available.
+
+## Outstanding Questions
+
+- What is an acceptable performance overhead for enabling tracing in user environments?
+- Are component-focused spans acceptable, or are there cases where event-focused spans would be clearly superior?
+- Are we comfortable with the proposed methods of avoiding the previous memory growth issue?
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues
+after the RFC is approved:
+
+- [ ] Adjust existing spike-level implementation to fully match what's described here
+- [ ] Instrument enough components to fully cover a chosen soak test scenario
+- [ ] Enable tracing in that soak test scenario, gathering data on performance impact
+- [ ] Open issues to add spans to the remaining components, organize and prioritize implementation
+
+Note: The final step is not necessarily required, and we could take the stance
+that this kind of instrumentation can happen as it is needed.
+
+## Future Improvements
+
+- Document specific performance impact of tracing based on measurements in our soak testing infrastructure
+- Improve observability of trace shipping infrastructure itself
+- Add more granular but more expensive span data at log levels like `DEBUG`
+- Enable sampling and options helping us towards a potential always-on system
+- Remove requirement of running a local Datadog agent
```

Comment:
As above I think this presumes that we're moving forward with tracing. FWIW my appetite is for a bounded experiment to determine how much of a game changer tracing is compared to our existing observability methods -- especially if we improve those -- to help decide whether we ought to move forward with it or pare back / remove it from the codebase. This RFC does mention that we're slightly outside of the normal request/response model for tracing.